Sia Integration

dvstate opened the pull request:
## Description
This code change adds a Sia gateway to Minio server, allowing data to be stored on the Sia network using Minio as a frontend. It changes nothing functional with the existing Minio codebase. It simply adds an additional gateway to allow users to store content on Sia.

Documentation for using the Sia gateway, which can be incorporated into Minio's documentation, can be found here:
https://github.com/dvstate/minio/blob/master/README.md

## Motivation and Context
To allow Minio users to utilize Sia for inexpensive backend storage.

## How Has This Been Tested?
New methods were individually tested as they were written.
Basic Go tests have been written for some of the Sia functionality.
Live servers using the Sia gateway have been operational for a couple weeks now, and all reported bugs, issues, and annoyances have been addressed.

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

david-sis committed:
All requirements stated in Sia bounty (issue 2155) have been met. Code is messy and not optimal in some areas, but can be quickly fixed. Rushing to commit to claim bounty. Will improve and support.

david-sis committed:
All requirements stated in Sia bounty (issue 2155) have been met. Code is messy and not optimal in some areas, but can be quickly fixed. Rushing to commit to claim bounty. Will improve and support.

david-sis committed:
Simplified sia integration in minio by creating a more general-purpose SiaBridge package at github.com/dvstate/siabridge

david-sis committed:
Simplified sia integration in minio by creating a more general-purpose SiaBridge package at github.com/dvstate/siabridge

david-sis committed:
Better filesystem management. Removed SiaBridge dep. ENVs for tunables. Better cache management.

david-sis committed:
Better filesystem management. Removed SiaBridge dep. ENVs for tunables. Better cache management.

david-sis committed:
More robust uploading. More descriptive error messages for cache issues.

david-sis committed:
More robust uploading. More descriptive error messages for cache issues.

david-sis committed:
Bugfix for some OSes

david-sis committed:
Bugfix for some OSes

david-sis committed:
Adding optional debug output

david-sis committed:
Adding optional debug output

david-sis committed:
Minor debug change

david-sis committed:
Minor debug change

david-sis committed:
Test before you commit, dummy

david-sis committed:
Test before you commit, dummy

david-sis committed:
Only check cachelayer on first chunk of GetObject call

david-sis committed:
Only check cachelayer on first chunk of GetObject call

david-sis committed:
Bug fix with GetObjectInfo after purging cache items. Partial error message improvement.

david-sis committed:
Bug fix with GetObjectInfo after purging cache items. Partial error message improvement.

david-sis committed:
Much improved error handling via Minio Browser, several stability improvements.

david-sis committed:
Much improved error handling via Minio Browser, several stability improvements.

david-sis committed:
Not much.

david-sis committed:
Not much.

david-sis committed:
Purge cache improvement

david-sis committed:
Purge cache improvement

david-sis committed:
Some debug info.

david-sis committed:
Some debug info.

david-sis committed:
Experiment...

david-sis committed:
Experiment...

david-sis committed:
URL encoding of filenames on Sia

david-sis committed:
URL encoding of filenames on Sia

david-sis committed:
Fixed create bucket bug I had introduced by changing error types.

david-sis committed:
Fixed create bucket bug I had introduced by changing error types.

david-sis committed:
Fixed some database concurrency bugs. Added bucket policy support.

david-sis committed:
Fixed some database concurrency bugs. Added bucket policy support.

david-sis committed:
Better treatment of special characters in filenames. Fixed a missing defer statement in markObjectCached.

david-sis committed:
Better treatment of special characters in filenames. Fixed a missing defer statement in markObjectCached.

david-sis committed:
Formatted using gofmt. Eliminated global variables.

david-sis committed:
Formatted using gofmt. Eliminated global variables.

david-sis committed:
Minor code refactor

david-sis committed:
Minor code refactor

harshavardhana committed:
Fix bucket policies support for gateway

harshavardhana committed:
Fix bucket policies support for gateway

david-sis committed:
Applying Minio-provided patch to fix gateway bucket policies.

david-sis committed:
Implemented anonymous functions in Sia Minio gateway.

david-sis committed:
Implemented anonymous functions in Sia Minio gateway.

david-sis committed:
Only print ENV values in debug mode. Added initial set of tests for cache layer.

david-sis committed:
Only print ENV values in debug mode. Added initial set of tests for cache layer.

david-sis committed:
Refactored return types in gateway-sia-cachelayer.go

david-sis committed:
Refactored return types in gateway-sia-cachelayer.go

david-sis committed:
Mistake in last commit. Updated tests.

david-sis committed:
Mistake in last commit. Updated tests.

david-sis committed:
Linted

david-sis committed:
Linted

david-sis committed:
Fixed typo

david-sis committed:
Fixed typo

harshavardhana commented on cmd/gateway-sia-anonymous.go:
```diff
@@ -0,0 +1,44 @@
+package cmd
```
Missing license header..

dvstate replied:
Included in latest commit. If license needs changed, just let me know.

harshavardhana replied:
Yes @dvstate it needs to say `Minio Cloud Storage` and not `David Gore`.  Will check back with @abperiasamy on this for final call.

dvstate replied:
Fixed in most recent commit.

harshavardhana commented on cmd/gateway-sia-cachelayer.go:
```diff
@@ -0,0 +1,1194 @@
+package cmd
```
License header..

dvstate replied:
Included in latest commit. If license needs changed, just let me know.

harshavardhana commented on cmd/gateway-sia-helpers.go:
```diff
@@ -0,0 +1,172 @@
+package cmd
```
License.. header.

dvstate replied:
Included in latest commit. If license needs changed, just let me know.

dvstate replied:
Fixed in most recent commit.

harshavardhana commented on cmd/server-main.go:
```diff
@@ -31,7 +31,7 @@ import (
 	cli.StringFlag{
 		Name:  "address",
-		Value: ":" + globalMinioPort,
+		Value: ":9000",
```
We should use globalMinioPort here

dvstate replied:
Change made in latest commit.

harshavardhana commented on cmd/gateway-sia_test.go:
```diff
@@ -0,0 +1,86 @@
+package cmd
```
License header..

dvstate replied:
Included in latest commit. If license needs changed, just let me know.

dvstate replied:
Fixed in most recent commit.

harshavardhana commented on cmd/gateway-sia-cachelayer.go:
```diff
@@ -0,0 +1,1194 @@
+func (cache *SiaCacheLayer) DeleteBucket(bucket string) *SiaServiceError {
+	cache.debugmsg("SiaCacheLayer.DeleteBucket")
+	// First, should delete all objects in the bucket
+	objects, err := cache.ListObjects(bucket)
```
Delete bucket shouldn't delete all top level objects.. only multipart i.e temporary objects can be deleted. Otherwise DeleteBucket() should return ENOTEMPTY.  Look at the fs-v1.go implementation for more inputs.

dvstate replied:
DeleteBucket should now behave as you described.

harshavardhana commented on cmd/gateway-sia-cachelayer.go:
```diff
@@ -0,0 +1,1194 @@
+	"encoding/json"
+	"fmt"
+	"github.com/NebulousLabs/Sia/api"
+	_ "github.com/mattn/go-sqlite3" // Has to be imported as blank
```
Lets actually explore on removing this dependency and moving to either https://github.com/pingcap/tidb or https://github.com/boltdb/bolt 

BoltDB is more native in its feel APIs are sort of synonymous and its quite small. 

https://github.com/pingcap/tidb is more feature rich and SQL support. 

I am leaning towards BoltDB but its your choice. We really need to get rid of CGO dependency.

dvstate replied:
The latest commit uses BoltDB and the CGO dependency has been eliminated.

harshavardhana reviewed (changes requested):
Initial comments.. Thanks for sending the PR @dvstate

vadmeste commented on cmd/gateway-sia-helpers.go:
```diff
@@ -0,0 +1,172 @@
+// abs returns the absolute representation of a path.
+// TODO: bad things can happen if you run siac from a non-existent directory.
+// Implement some checks to catch this problem.
+func abs(path string) string {
```
I am not sure why we use abs() here instead of using directly filepath.Abs()

dvstate replied:
Changes made as requested in latest commit.

vadmeste commented on cmd/gateway-sia.go:
```diff
@@ -0,0 +1,431 @@
+func newSiaGateway(host string) (GatewayLayer, error) {
+	sia := &siaObjects{
+		PurgeCacheAfterSec: 24 * 60 * 60,
+		SiadAddress:        "127.0.0.1:9980",
```
Do you mean `host` here instead of `127.0.0.1:9980` ?

dvstate replied:
Change made as requested so that the Sia daemon api address can be specified on command line.

david-sis committed:
Made changes requested in Minio PR, except for DB.

david-sis committed:
Made changes requested in Minio PR, except for DB.

david-sis committed:
Switched to BoltDB, improved organization, more thorough tests.

david-sis committed:
Switched to BoltDB, improved organization, more thorough tests.

david-sis committed:
Added SIA_BACKGROUND_UPLOAD env to allow users to optionally upload to Sia in background

david-sis committed:
Added SIA_BACKGROUND_UPLOAD env to allow users to optionally upload to Sia in background

david-sis committed:
Merge branch 'sia_integration_v3' into sia_integration_v2

david-sis committed:
More intuitive assignment of Sia daemon address via CLI, ENV, and default.

david-sis committed:
More intuitive assignment of Sia daemon address via CLI, ENV, and default.

david-sis committed:
Fixed cache space bug.

david-sis committed:
Fixed cache space bug.

harshavardhana commented on browser/app/js/actions.js:
```diff
@@ -492,6 +492,16 @@ export const uploadFile = (file, xhr) => {
         }))
       }
+      if (xhr.status == 500) {
+        setShowAbortModal(false)
```
what is this change for? looks like unrelated. Would it be possible to send it later after this PR is integrated into master? . Perhaps you can send this separately as well? @dvstate

dvstate replied:
The section of code added to actions.js is to allow the javascript alert dialog to display a descriptive and meaningful error message. Otherwise the user doesn't know specifically what the problem is. Quite a few users requested this ability during testing.

harshavardhana replied:
Can you send this change as separate PR? @dvstate do not think this belongs to this work..

harshavardhana commented on cmd/gateway-main.go:
```diff
@@ -125,6 +125,22 @@ EXAMPLES:
+  {{end}}{{end}}
+
+EXAMPLES:
+  1. Start minio gateway server for GCS backend.
```
Fix the GCS should say Sia backend here.. and perhaps also should list out an example setup here.

dvstate replied:
Corrected. Also added all ENV variables and descriptions and defaults to the printout message.

harshavardhana commented on cmd/gateway-handlers.go:
```diff
@@ -509,9 +512,13 @@ func (api gatewayAPIHandlers) DeleteBucketPolicyHandler(w http.ResponseWriter, r
+
+	globalBucketPolicies.SetBucketPolicy(bucket, policyChange{
+		true, policy.BucketAccessPolicy{},
+	})
```
I would suggest you to remove the bucket policy related changes we decided to do it differently and not keep/manage any in memory state for gateway code.  So you can perhaps just remove my PR in this regard.  We can say currently bucket policies do not work on Sia backend and fix it later.

dvstate replied:
So you want me to remove all the changes that you made to support bucket policies? There are multiple servers running now by Sia users that depend on bucket policy support. I guess they could continue using a branch of my version of Minio until policies are fully supported, but I do imagine you'll get quite a bit of feedback from users regarding policies not working. :)

harshavardhana replied:
Yes the problem is not that, the problem is that its not a correct approach and has lots of holes in it.  It is okay to loose this ability while we address the overall architecture.  I reached out to you on slack, perhaps we can discuss on other things in the PR.

harshavardhana commented on cmd/gateway-sia-cachelayer.go:
```diff
@@ -0,0 +1,685 @@
+import (
+	"encoding/json"
+	"fmt"
+	"github.com/boltdb/bolt"
```
Move the `github.com` imports as mentioned here - you can also use `goimports` to do this for you. https://github.com/golang/go/wiki/CodeReviewComments#imports

harshavardhana commented on cmd/gateway-sia-cachelayer.go:
```diff
@@ -0,0 +1,685 @@
+		}
+	}
+
+	err := cache.dbInsertObject(bucket, objectName, size, time.Now().Unix(), 0, purgeAfter, srcFile, 1)
```
~~Does this insert the entire object to `boltDB` ? if yes have you tested with large files? - Would it be possible to use boltDB just for namespace and use regular filesystem for staging? it would be faster..~~

harshavardhana replied:
I see that its just namespace..

dvstate replied:
Correct. DB is only used to store small metadata for buckets and objects.

harshavardhana commented on cmd/gateway-sia-cachelayer.go:
```diff
@@ -0,0 +1,685 @@
+}
+
+// InsertBucket will attempt to insert a new bucket
+func (cache *SiaCacheLayer) InsertBucket(bucket string) *SiaServiceError {
```
Should this be just `MakeBucket` ? why a new API?

harshavardhana commented on cmd/gateway-sia-db.go:
```diff
@@ -0,0 +1,843 @@
+	defer cache.dbUnlockDB()
+
+	// Start a transaction
+    tx, err := cache.Db.Begin(true)
```
Looks like some code formatting issues here..

harshavardhana reviewed (changes requested):
Some review comments.. testing right now.

david-sis committed:
Minor debug logging enhancement.

david-sis committed:
Minor debug logging enhancement.

david-sis committed:
Another round of modifications requested by Minio PR

david-sis committed:
Another round of modifications requested by Minio PR

david-sis committed:
UNTESTED/UNREADY: Committing and pushing for the sake of showing progress on PR. Complete, but needs a few more hours of testing...

david-sis committed:
UNTESTED/UNREADY: Committing and pushing for the sake of showing progress on PR. Complete, but needs a few more hours of testing...

david-sis committed:
Initial testing/verification complete.

david-sis committed:
Initial testing/verification complete.

david-sis committed:
Reverted .gitignore. Ran gofmt on code changes.

david-sis committed:
Reverted .gitignore. Ran gofmt on code changes.

tlightsky commented on cmd/gateway-sia.go:
```diff
@@ -23,17 +23,17 @@ import (
-	CacheDir           string         // Name of cache directory.
-	DbFile             string         // Name of cache database file.
+	CacheDir           string         // Temporary storage location for file transfers.
+	SiaRootDir         string         // Root directory to store files on Sia.
```
what is this `SiaRootDir` for, i think siad should be ready before start minio

tlightsky commented on cmd/gateway-main.go:
```diff
@@ -136,14 +136,9 @@ FLAGS:
-  SIA_MANAGER_DELAY_SEC:     The number of seconds to delay between cache/db management operations. (30)
-  SIA_UPLOAD_CHECK_FREQ_MS:  The time in milliseconds to wait between checks with Sia network to determine if file has completed uploading. (3000)
-  SIA_CACHE_MAX_SIZE_BYTES:  The maximum allowed size of the cache directory in bytes. (10000000000)
-  SIA_CACHE_PURGE_AFTER_SEC: The number of seconds since the time a file was last downloaded before removing that file from the cache. (86400)
```
how long will the cached files purged then?

dvstate replied:
All cache functionality has been removed.

harshavardhana committed:
Change a whole bunch of APIs, simplify.

Remove dependency on FS object layer, SIA password
setup is necessary through ENV since signature v4
is still needed as form of access.

Removed anonymous support for now, we can bring it
again as needed.

harshavardhana committed:
Add dependencies

harshavardhana committed:
Revert "Fix bucket policies support for gateway"

This reverts commit 7cfa61d687277b4bc172b665aa66f067ba1a59c6.

david-sis committed:
Pulling in Harshavardhana's changes and pulling in latest Minio code. Tested to work with Sia.

david-sis committed:
Merge branch 'harshavardhana-dvstate-sia_integration_v2' into sia_integration_v2

mtlynch commented on cmd/gateway-sia.go:
```diff
@@ -0,0 +1,525 @@
+	// Need to wait for upload to complete
+	s.waitTillSiaUploadCompletes(bucket, object)
+
+	return objInfo, nil
```
Won't `objInfo` still be uninitialized?

harshavardhana replied:
No

david-sis committed:
Changed 'CacheDir' references to 'TempDir'. Default TempDir is '.sia_temp'. Updated isValidObjectName() regexp.

mtlynch commented on cmd/gateway-sia.go:
```diff
@@ -227,10 +227,15 @@ func newSiaGateway() (GatewayLayer, error) {
 
+	// If local Sia temp directory not specified, default to:
+	if sia.TempDir == "" {
+		sia.TempDir = ".sia_temp"
```
Can we add this to the `.gitignore`?

harshavardhana committed:
Merge branch 'master' into sia_integration_v2
